workflow/scripts/osemosys_global/powerplant/renewable_targets.py

```diff
@@ -0,0 +1,212 @@
+                sp_demand_df = annual_demand_df.loc[
+                    ~(annual_demand_df["FUEL"].str[3:8].isin(remove_nodes))
+                ]
+                  
+                if len(str(node_dict[target])) == 3:
+                    sp_demand_df["GEO"] = sp_demand_df["FUEL"].str[3:6]
+                    
+                elif len(str(node_dict[target])) == 5:
+                    sp_demand_df["GEO"] = sp_demand_df["FUEL"].str[3:8]
+                    
+                sp_demand_df = sp_demand_df.groupby(["YEAR", "GEO"], as_index=False)[
+                    "VALUE"
+                ].sum()
+                sp_demand_df.rename(columns={"VALUE": "DEMAND"}, inplace=True)
+                
+                # Merge RE targets and SpecifiedAnnualDemand tables
+                re_targets_df = pd.merge(
+                    re_targets_df, sp_demand_df, how="left", on=["GEO", "YEAR"]
+                )
+                
+                # Calculate the country level target for the dummy RES commodity
+                re_targets_df["VALUE"] = re_targets_df["VALUE"] * re_targets_df["DEMAND"]
+                re_targets_df["FUEL"] = target + re_targets_df["GEO"]
+                re_targets_df["REGION"] = region_name
+                re_targets_df = re_targets_df[["REGION", "FUEL", "YEAR", "VALUE"]]
+                
+                if accumulated_annual_demand.empty:
+                
+                    accumulated_annual_demand = re_targets_df
+                
+                else:
+                    accumulated_annual_demand = pd.concat([
+                        accumulated_annual_demand, 
+                        re_targets_df])
+
+    return fuel_set, oar_df, accumulated_annual_demand
+
+def apply_re_abs_targets(re_targets, remove_nodes, 
+                         tech_set, region_name):
+    
+    """Apply Renewable Energy targets by country, year and technology
+    in absolute terms (GW)."""
+    
+    if not remove_nodes:
+        remove_nodes = []
+        
+    node_dict = {}
+    target_techs_dict = {}
+    type_dict = {}
+    first_year_dict = {}
+    final_year_dict = {}
+    value_dict = {}
+        
+    total_annual_min_capacity = pd.DataFrame(columns = ["REGION", "TECHNOLOGY", 
+                                                        "YEAR", "VALUE"])
+    # Check if any targets are defined.
+    if not re_targets is None:
+        # Get transmission parameters from config file.
+        for target, target_params in re_targets.items():
+            node_dict[target] = target_params[0]
+            target_techs_dict[target] = target_params[1]
+            type_dict[target] = target_params[2]
+            first_year_dict[target] = target_params[3]
+            final_year_dict[target] = target_params[4]
+            value_dict[target] = target_params[5]
+
+        # Per entry check if they are absolute targets ('ABS').
+        for target in re_targets.keys():
```

Comment:
Not a required change, but you can drop the `.keys()` call, as `in` will only loop over the keys. ie. 

```
for target in re_targets:
  ...
```